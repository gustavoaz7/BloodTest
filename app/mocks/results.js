export const resultsMock = [
  {
    id: '0b82fed2-a091-4318-ba02-4d5e6192d6f7',
    name: 'Citlalli Kuhic',
    email: 'Berniece.Haley42@hotmail.com',
    samples: [
      {
        time: 'morning',
        value: 0.48,
      },
      {
        time: 'afternoon',
        value: 0.39,
      },
      {
        time: 'evening',
        value: 0.19,
      },
    ],
  },
  {
    id: 'acf65e6a-1878-4cd9-bf6f-440dafd4fa18',
    name: 'Rachael Lehner',
    email: 'Teagan_Ratke83@gmail.com',
    samples: [
      {
        time: 'morning',
        value: 0.67,
      },
      {
        time: 'afternoon',
        value: 0.31,
      },
      {
        time: 'evening',
        value: 0.24,
      },
    ],
  },
  {
    id: 'ca285be1-0d6b-42f6-bfec-b3aeaa2aaf8f',
    name: 'Willa Quitzon',
    email: 'Palma17@yahoo.com',
    samples: [
      {
        time: 'morning',
        value: 0.73,
      },
      {
        time: 'afternoon',
        value: 0.19,
      },
      {
        time: 'evening',
        value: 0.14,
      },
    ],
  },
  {
    id: '95eba822-a7b3-4938-8c94-005733701c1f',
    name: 'Miss Audie Moore',
    email: 'Giuseppe19@hotmail.com',
    samples: [
      {
        time: 'morning',
        value: 0.26,
      },
      {
        time: 'afternoon',
        value: 0.09,
      },
      {
        time: 'evening',
        value: 0.29,
      },
    ],
  },
  {
    id: '4fea89aa-bcf7-42b0-a5e8-202fc03a02fe',
    name: 'Yesenia Nicolas',
    email: 'Maeve_Jacobs14@yahoo.com',
    samples: [
      {
        time: 'morning',
        value: 0.78,
      },
      {
        time: 'afternoon',
        value: 0.26,
      },
      {
        time: 'evening',
        value: 0.24,
      },
    ],
  },
  {
    id: '17d11f05-2580-49b5-bb17-b786209c4792',
    name: 'Ms. Harmon Kunze',
    email: 'Johnson90@yahoo.com',
    samples: [
      {
        time: 'morning',
        value: 0.47,
      },
      {
        time: 'afternoon',
        value: 0.09,
      },
      {
        time: 'evening',
        value: 0.09,
      },
    ],
  },
];
