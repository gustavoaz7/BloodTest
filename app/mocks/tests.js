export const testsMock = [
  {
    id: '5d41b693ed44f7ee9f9436d0',
    title: 'Healthcare',
    date: '17.07.2019',
    result: [
      {
        biomarker: 'fats',
        color: '#e30613',
        percentage: 25,
      },
      {
        biomarker: 'carbohydrates',
        color: '#D3235D',
        percentage: 25,
      },
      {
        biomarker: 'vitamins',
        color: '#547F81',
        percentage: 25,
      },
      {
        biomarker: 'magnesium',
        color: '#0085FF',
        percentage: 25,
      },
    ],
  },
  {
    id: '5d41b6938f1360cdde5ecc89',
    title: 'Advanced',
    date: '08.07.2019',
    result: [
      {
        biomarker: 'protein',
        color: '#E2D669',
        percentage: 25,
      },
      {
        biomarker: 'vitamins',
        color: '#ffcc00',
        percentage: 25,
      },
      {
        biomarker: 'magnesium',
        color: '#e30613',
        percentage: 25,
      },
      {
        biomarker: 'carbohydrates',
        color: '#0085FF',
        percentage: 25,
      },
    ],
  },
  {
    id: '5d41b693084f4f55217fd864',
    title: 'Sport',
    date: '30.07.2019',
    result: [
      {
        biomarker: 'fiber',
        color: '#E2D669',
        percentage: 25,
      },
      {
        biomarker: 'fats',
        color: '#D3235D',
        percentage: 25,
      },
      {
        biomarker: 'magnesium',
        color: '#547F81',
        percentage: 25,
      },
      {
        biomarker: 'vitamins',
        color: '#0085FF',
        percentage: 25,
      },
    ],
  },
];
