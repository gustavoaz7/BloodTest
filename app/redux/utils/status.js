import { ActionType } from 'redux-promise-middleware';

export const stateWithStatus = {
  value: null,
  status: null,
};

export function setPending(state) {
  return { ...state, status: ActionType.Pending };
}

export function setFulfilled(state, payload = null) {
  return { ...state, status: ActionType.Fulfilled, value: payload };
}

export function setRejected(state, payload = null) {
  return { ...state, status: ActionType.Rejected, value: payload };
}

export function isPending(value) {
  return value.status === ActionType.Pending;
}

export function isFulfilled(value) {
  return value.status === ActionType.Fulfilled;
}

export function isRejected(value) {
  return value.status === ActionType.Rejected;
}

export function isInitialized(value) {
  return Boolean(value.status);
}
