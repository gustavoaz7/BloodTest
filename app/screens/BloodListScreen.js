import React, { Component } from 'react';
import { View, FlatList, Alert, StyleSheet } from 'react-native';
import styled from 'styled-components/native';
import Loading from '../components/Loading';
import ListItem from '../components/ListItem';
import { getBloodList } from '../services/api';

export default class BloodListScreen extends Component {
  constructor(props) {
    super(props);
    this.state = {
      tests: [],
      error: false,
      loading: true,
      refreshing: false,
    };
  }

  componentDidMount() {
    this.loadTests();
  }

  loadTests = async () => {
    try {
      const tests = await getBloodList();
      const sortedTests = this.sortTests(tests);
      this.setState({ tests: sortedTests, loading: false, error: false, refreshing: false });
    } catch (e) {
      this.setState({ tests: [], error: true, loading: false, refreshing: false });
      console.warn('BloodListScreen load tests error - ', e);
    }
  };

  handleRefresh = () => {
    this.setState({ refreshing: true });
    this.loadTests();
  };

  sortTests(tests) {
    return tests.sort((a, b) => {
      const aa = a.date
        .split('.')
        .reverse()
        .join();
      const bb = b.date
        .split('.')
        .reverse()
        .join();
      if (aa < bb) return 1;
      if (aa > bb) return -1;
      return 0;
    });
  }

  handleError() {
    Alert.alert(
      'Something went wrong :(',
      'There was a problem connecting to our servers.\n\nPlease try again in a few seconds.',
      [
        {
          text: 'Try again',
          onPress: this.loadTests,
        },
      ],
    );
  }

  keyExtractor(item) {
    return item.id;
  }

  renderItem({ item }) {
    return <ListItem item={item} />;
  }

  renderSeparator() {
    return <Separator />;
  }

  render() {
    const { tests, loading, error, refreshing } = this.state;
    if (loading) return <Loading />;
    if (error) {
      this.handleError();
      return null;
    }
    return (
      <FlatListUI
        data={tests}
        renderItem={this.renderItem}
        keyExtractor={this.keyExtractor}
        ItemSeparatorComponent={this.renderSeparator}
        refreshing={refreshing}
        onRefresh={this.handleRefresh}
        contentContainerStyle={styles.contentStyle}
      />
    );
  }
}

export const FlatListUI = styled(FlatList)`
  background-color: 'rgba(0, 0, 0, 0.05)';
`;

const Separator = styled(View)`
  height: 12px;
`;

const styles = StyleSheet.create({
  contentStyle: {
    padding: 12,
  },
});
