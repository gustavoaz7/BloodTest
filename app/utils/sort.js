export const sortTestsByDate = tests =>
  tests.sort((a, b) => {
    const aa = a.date
      .split('.')
      .reverse()
      .join();
    const bb = b.date
      .split('.')
      .reverse()
      .join();
    if (aa < bb) return 1;
    if (aa > bb) return -1;
    return 0;
  });
