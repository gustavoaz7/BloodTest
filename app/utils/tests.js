export const flushPromises = () => new Promise(setImmediate);

export const stackNavigationProps = {
  state: {},
  dispatch() {},
  goBack() {},
  dismiss() {},
  navigate() {},
  getParam() {},
  setParams() {},
  emit() {},
  addListener() {},
  isFocused() {},
  isFirstRouteInParent() {},
  dangerouslyGetParent() {},
  router() {},
};
