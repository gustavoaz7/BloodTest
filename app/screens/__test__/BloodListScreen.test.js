import React from 'react';
import { shallow } from 'enzyme';
import axios from 'axios';
import BloodListScreen, { FlatListUI } from '../BloodListScreen';
import { testsMock } from '../../mocks/tests';
import { flushPromises } from '../../utils/tests';

const sortedTestsMock = [testsMock[2], testsMock[0], testsMock[1]];

describe('BloodListScreen', () => {
  const testsResponse = { data: testsMock };
  const axiosMock = jest.fn().mockImplementation(() => Promise.resolve(testsResponse));

  beforeEach(() => {
    axiosMock.mockClear();
    axios.get = axiosMock;
  });

  it('renders without crashing', () => {
    shallow(<BloodListScreen />);
  });

  it('sorts blood tests correctly', () => {
    expect(BloodListScreen.prototype.sortTests(testsMock)).toEqual(sortedTestsMock);
  });

  it('loads sorted blood tests when mouting', async () => {
    const wrapper = shallow(<BloodListScreen />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'sortTests');

    await flushPromises();

    expect(axiosMock).toHaveBeenCalled();
    expect(instance.sortTests).toHaveBeenCalled();
    expect(wrapper.state('tests')).toEqual(sortedTestsMock);
  });

  it('loads sorted blood tests on `pull to refresh` ', async () => {
    const wrapper = shallow(<BloodListScreen />);
    const instance = wrapper.instance();
    const spy = jest.spyOn(instance, 'loadTests');

    wrapper.update();
    instance.forceUpdate();

    await flushPromises();

    wrapper
      .find(FlatListUI)
      .props()
      .onRefresh();

    await flushPromises();

    expect(spy).toHaveBeenCalledTimes(1);
  });

  describe('matches snapshot', () => {
    it('loading state', () => {
      const wrapper = shallow(<BloodListScreen />);

      expect(wrapper).toMatchSnapshot('loading');
    });

    it('error state', async () => {
      const mock = jest.fn().mockImplementation(() => {
        throw new Error('error');
      });
      axios.get = mock;
      const wrapper = shallow(<BloodListScreen />);
      await flushPromises();

      expect(wrapper).toMatchSnapshot('error');
    });

    it('success state', async () => {
      const wrapper = shallow(<BloodListScreen />);
      await flushPromises();

      expect(wrapper).toMatchSnapshot('success');
    });
  });
});
