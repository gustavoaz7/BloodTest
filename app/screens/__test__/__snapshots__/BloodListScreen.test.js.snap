// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BloodListScreen matches snapshot pending state: pending 1`] = `
<PendingWrapper
  loadingProps={null}
  value={
    Object {
      "status": "PENDING",
      "value": null,
    }
  }
>
  <Styled(FlatList)
    ItemSeparatorComponent={[Function]}
    contentContainerStyle={
      Object {
        "padding": 12,
      }
    }
    data={null}
    keyExtractor={[Function]}
    onRefresh={[Function]}
    refreshing={false}
    renderItem={[Function]}
  />
</PendingWrapper>
`;

exports[`BloodListScreen matches snapshot rejected state: rejected 1`] = `""`;

exports[`BloodListScreen matches snapshot success state: success 1`] = `
<PendingWrapper
  loadingProps={null}
  value={
    Object {
      "status": "FULFILLED",
      "value": Array [
        Object {
          "date": "17.07.2019",
          "id": "5d41b693ed44f7ee9f9436d0",
          "result": Array [
            Object {
              "biomarker": "fats",
              "color": "#e30613",
              "percentage": 25,
            },
            Object {
              "biomarker": "carbohydrates",
              "color": "#D3235D",
              "percentage": 25,
            },
            Object {
              "biomarker": "vitamins",
              "color": "#547F81",
              "percentage": 25,
            },
            Object {
              "biomarker": "magnesium",
              "color": "#0085FF",
              "percentage": 25,
            },
          ],
          "title": "Healthcare",
        },
        Object {
          "date": "08.07.2019",
          "id": "5d41b6938f1360cdde5ecc89",
          "result": Array [
            Object {
              "biomarker": "protein",
              "color": "#E2D669",
              "percentage": 25,
            },
            Object {
              "biomarker": "vitamins",
              "color": "#ffcc00",
              "percentage": 25,
            },
            Object {
              "biomarker": "magnesium",
              "color": "#e30613",
              "percentage": 25,
            },
            Object {
              "biomarker": "carbohydrates",
              "color": "#0085FF",
              "percentage": 25,
            },
          ],
          "title": "Advanced",
        },
        Object {
          "date": "30.07.2019",
          "id": "5d41b693084f4f55217fd864",
          "result": Array [
            Object {
              "biomarker": "fiber",
              "color": "#E2D669",
              "percentage": 25,
            },
            Object {
              "biomarker": "fats",
              "color": "#D3235D",
              "percentage": 25,
            },
            Object {
              "biomarker": "magnesium",
              "color": "#547F81",
              "percentage": 25,
            },
            Object {
              "biomarker": "vitamins",
              "color": "#0085FF",
              "percentage": 25,
            },
          ],
          "title": "Sport",
        },
      ],
    }
  }
>
  <Styled(FlatList)
    ItemSeparatorComponent={[Function]}
    contentContainerStyle={
      Object {
        "padding": 12,
      }
    }
    data={
      Array [
        Object {
          "date": "17.07.2019",
          "id": "5d41b693ed44f7ee9f9436d0",
          "result": Array [
            Object {
              "biomarker": "fats",
              "color": "#e30613",
              "percentage": 25,
            },
            Object {
              "biomarker": "carbohydrates",
              "color": "#D3235D",
              "percentage": 25,
            },
            Object {
              "biomarker": "vitamins",
              "color": "#547F81",
              "percentage": 25,
            },
            Object {
              "biomarker": "magnesium",
              "color": "#0085FF",
              "percentage": 25,
            },
          ],
          "title": "Healthcare",
        },
        Object {
          "date": "08.07.2019",
          "id": "5d41b6938f1360cdde5ecc89",
          "result": Array [
            Object {
              "biomarker": "protein",
              "color": "#E2D669",
              "percentage": 25,
            },
            Object {
              "biomarker": "vitamins",
              "color": "#ffcc00",
              "percentage": 25,
            },
            Object {
              "biomarker": "magnesium",
              "color": "#e30613",
              "percentage": 25,
            },
            Object {
              "biomarker": "carbohydrates",
              "color": "#0085FF",
              "percentage": 25,
            },
          ],
          "title": "Advanced",
        },
        Object {
          "date": "30.07.2019",
          "id": "5d41b693084f4f55217fd864",
          "result": Array [
            Object {
              "biomarker": "fiber",
              "color": "#E2D669",
              "percentage": 25,
            },
            Object {
              "biomarker": "fats",
              "color": "#D3235D",
              "percentage": 25,
            },
            Object {
              "biomarker": "magnesium",
              "color": "#547F81",
              "percentage": 25,
            },
            Object {
              "biomarker": "vitamins",
              "color": "#0085FF",
              "percentage": 25,
            },
          ],
          "title": "Sport",
        },
      ]
    }
    keyExtractor={[Function]}
    onRefresh={[Function]}
    refreshing={false}
    renderItem={[Function]}
  />
</PendingWrapper>
`;
